hom/utils.py

```diff
@@ -0,0 +1,267 @@
+    )
+
+
+def get_channel_by_id(
+    bot: commands.Bot, guild_id: int, channel_id: int, channel_type: str = "text"
+) -> t.Optional[discord.TextChannel]:
+    if guild_id not in [guild.id for guild in bot.guilds]:
+        return None
+
+    if not (guild := bot.get_guild(guild_id)):
+        return None
+
+    channel = discord.utils.find(
+        lambda c: c.id == channel_id and str(c.type).lower() == channel_type,
+        guild.channels,
+    )
+
+    return t.cast(discord.TextChannel, channel) if channel else None
+
+
+def get_channel_by_name(
+    bot: commands.Bot | discord.Client,
+    guild_id: int,
+    channel_name: str,
+    channel_type: str = "text",
+) -> t.Optional[discord.TextChannel]:
+    if not any(guild_id == guild.id for guild in bot.guilds):
+        return None
+
+    if not (guild := bot.get_guild(guild_id)):
+        return None
+
+    channel = discord.utils.find(
+        lambda c: normalize(c.name) == normalize(channel_name)
+        and str(c.type).lower() == channel_type,
+        guild.channels,
+    )
+
+    return t.cast(discord.TextChannel, channel) if channel else None
+
+
+async def get_user_by_original_message(
+    channel: discord.TextChannel,
+) -> t.Optional[discord.Member | discord.User]:
+    messages = [message async for message in channel.history(limit=1, oldest_first=True)]
+    mentions = messages[0].mentions if messages else None
+    return mentions[0] if mentions else None
+
+
+def get_user_ticket_channel(
+    guild: discord.Guild, user: discord.User | discord.Member
+) -> t.Optional[discord.TextChannel]:
+    if category := get_category(guild, "Tickets"):  # TODO: Make this channel_id?
+        for channel in category.channels:
+            user_perms = channel.overwrites_for(user)
+
+            if user_perms.view_channel:
+                return t.cast(discord.TextChannel, channel)
+
+    return None
+
+
+async def create_ticket_for_user(
+    interaction: discord.Interaction[commands.Bot],
+    instructions: str,
+    button_label: t.Optional[str],
+    example_url: t.Optional[str] = None,
+) -> discord.TextChannel:
+    assert interaction.guild
+    assert isinstance(interaction.user, discord.Member)
+    existing_ticket_channel = get_user_ticket_channel(interaction.guild, interaction.user)
+
+    if existing_ticket_channel:
+        msg_content = f":envelope:  Click [here]({existing_ticket_channel.jump_url}) to view your open ticket."
+        await interaction.followup.send(content=msg_content, ephemeral=True)
+        return existing_ticket_channel
+
+    channel_name = f"help-{interaction.user.display_name[:15]}"
+    tickets_category = get_category(interaction.guild, "Tickets")  # TODO: Make this channel_id?
+    mod_role = get_role_by_name(interaction.guild, "Moderator")
+    helpful_old_man_role = get_role_by_name(interaction.guild, "Helpful Old Man")
+    assert helpful_old_man_role
+    assert mod_role
```

Comment:
These asserts will fail unless I have my roles set to `Moderator` and `Helpful Old Man`. Maybe it would be better to use role IDs instead and define them in `.env`.

Reply:
yes, role ids will be better but the asserts are still necessary. because those roles need to be there for the bot to function.